1.21.4 Resource Pack Support

SpecialBuilder32 opened the pull request:
Adds 1.21.4. item model definition files to the resource pack, and moves existing models to an overlay for backwards compatibility. 

Remaining tasks:
- [ ] Testing in-game
- [x] RP marked valid versions (1.21.3 should not throw warning)
- [x] Restore backwards compatibility for shamir textures
- [ ] Fix vanilla item-model forwarding

SpecialBuilder32 committed:
generate new item model definitions

SpecialBuilder32 committed:
item-model merge policy

SpecialBuilder32 committed:
Generate 1.21.3 RP models into an overlay

SpecialBuilder32 committed:
Add CMD range padding into new item_def files

misode reviewed (changes requested):
I'm getting lots of resource pack load errors.
![image](https://github.com/user-attachments/assets/4b835050-fb42-4b1f-b2cf-dafa5a21073b)
![image](https://github.com/user-attachments/assets/0ef25710-7ce5-4cbf-a356-bf453dd0db4c)

SpecialBuilder32 commented:
> I'm getting lots of resource pack load errors. ![image](https://private-user-images.githubusercontent.com/17352009/418374969-4b835050-fb42-4b1f-b2cf-dafa5a21073b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA5NDY1OTksIm5iZiI6MTc0MDk0NjI5OSwicGF0aCI6Ii8xNzM1MjAwOS80MTgzNzQ5NjktNGI4MzUwNTAtZmI0Mi00YjFmLWIyY2YtZGFmYTVhMjEwNzNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzAyVDIwMTEzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjN2U5ODlhNDcyZGYwNTFjZDUyMmI3OWU1MDAwYWE0ZGVhNzhhM2JhZDc0Nzg3YmU3YWQ1OGEwNDJhOTY4NzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WXT1_385M61mZ2cGJmpK1UZX_CT4H3cI-qDArNNOGdQ) ![image](https://private-user-images.githubusercontent.com/17352009/418375009-0ef25710-7ce5-4cbf-a356-bf453dd0db4c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA5NDY1OTksIm5iZiI6MTc0MDk0NjI5OSwicGF0aCI6Ii8xNzM1MjAwOS80MTgzNzUwMDktMGVmMjU3MTAtN2NlNS00Y2JmLWEzNTYtYmY0NTNkZDBkYjRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzAyVDIwMTEzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmNzM4MDE2Mjk0YTI1ODJmODE2ODg1MjdlYTM5MDIxOTI3NDhmMjc3NTAyNTk2YjYwNjExMWIyZmJiZjgwMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gNECJhmWa91YlzZnp2Uwx9KDDWaWW1hGVlM0dTNJvSw)

I was almost certain it did support item-model-definition compounds both when I tested it and re the wiki. 
![image](https://github.com/user-attachments/assets/e57b92bf-aa49-49da-ae0b-93ce5eb91bf7)
I'll do some more digging and see why I didn't get this problem on my end and/or what has changed

misode commented:
@SpecialBuilder32 Have you had time to look further into these issues? Given that 1.21.5-pre1 just released, it would be ideal if this PR is ready to be merged before we have to worry with the 1.21.5 update.

SpecialBuilder32 committed:
Replace manual predicate-entry into model data config (for end fishing elytra)

SpecialBuilder32 committed:
Manually provide end fishing elytra item model in overlay

SpecialBuilder32 commented:
In tracking down what the load errors were, I realized there were some special cases I needed to manually handle. I had *thought* it was just the end fishing elytra but now that I see build errors I realize armor trims must also be redone :/

SpecialBuilder32 committed:
Update shamir model/texture generation code for 1.21.4

SpecialBuilder32 committed:
Add backwards compatibility for 1.21.3 metallurgy textures

SpecialBuilder32 committed:
Fix references to item references that only exist as block references

SpecialBuilder32 committed:
Add support for special-case vanilla tempates

Covers player heads, which is used by metallurgy mould heads. NOTE: current implementation puts heads on an armor stand, bypassing the usual model rendering. If a custom block model is added to the moulds, migration to an item-display is required.

SpecialBuilder32 committed:
Add rename for guidebook broken elytra texture

SpecialBuilder32 committed:
[Incomplete] fixing missing advancement icon model forwarding

SpecialBuilder32 committed:
[Incomplete] Move item-def handling to Templates

SpecialBuilder32 commented:
Alright just a quick update on this - after running into too many edge cases, I'm half-rewriting the 1.20.4/1.20.5 update to make more sense and handle model-file/item-definition interactions better. It'll be a fair amount of more testing and debugging until this is ready. Sorry 😢

SpecialBuilder32 committed:
VanillaTemplate pulls in default model settings

SpecialBuilder32 committed:
[Incomplete] progress from desktop

SpecialBuilder32 committed:
Advancement models inherit from VanillaTemplate

SpecialBuilder32 committed:
Fix shield models being invisible

SpecialBuilder32 committed:
Sunken treasure chests and metallurgy elytra

misode commented:
@SpecialBuilder32 With #1107 being merged now, is this PR also good to go?

Bloo-dev commented:
For future people reading this: RP support for 1.21.4 is still missing as of today.

Bloo-dev commented:
Due to our limited time budget, we decided not to push this PR.
